File: lesson_5_1/app/src/main/java/com/kalashnyk/denys/simplemvpapp/presentation/base/BaseAdapter.java
package com.kalashnyk.denys.simplemvpapp.presentation.base;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;

import java.util.List;

public abstract class BaseAdapter<VH extends RecyclerView.ViewHolder, M, L> extends RecyclerView.Adapter<VH> {

    protected List<M> mList;
    protected Context mContext;
    protected L mItemClickListener;

    public BaseAdapter(Context context, List<M> list) {
        mList = list;
        mContext = context;
    }

    @Override
    public void onBindViewHolder(@NonNull VH holder, int position) {
        holder.itemView.setTag(mList.get(position));
    }

    public M getItem(int position) {
        return mList.get(position);
    }

    public int getItemCount() {
        return mList.size();
    }

    public void add(M item) {
        mList.add(item);
    }

    public void addAll(List<M> list) {
        for (M item : list) {
            add(item);
        }
    }

    public void remove(M item) {
        int position = mList.indexOf(item);
        if (position > -1) {
            mList.remove(position);
            notifyItemRemoved(position);
        }
    }

    public void clear() {
        mList.clear();
        notifyDataSetChanged();
    }

    public void setItemClickListener(L listener){
        mItemClickListener = listener;
    }
}

File: lesson_9/app/src/main/java/com/kalashnyk/denys/kotlinbindingsample/presentation/base/BaseAdapter.kt
package com.kalashnyk.denys.kotlinbindingsample.presentation.base

import android.support.v7.widget.RecyclerView

abstract class BaseAdapter<VH : RecyclerView.ViewHolder, M, L>
    (protected var list: MutableList<M>, protected var clickListener: L) : RecyclerView.Adapter<VH>() {

    override fun onBindViewHolder(holder: VH, position: Int) {
        holder.itemView.tag = list[position]
    }

    fun getItem(position: Int): M {
        return list[position]
    }

    override fun getItemCount(): Int {
        return list.size
    }

    fun add(item: M) {
        list.add(item)
    }

    fun addAll(list: List<M>) {
        for (item in list) {
            add(item)
        }
    }

    fun remove(item: M) {
        val position = list.indexOf(item)
        if (position > -1) {
            list.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    fun clear() {
        list.clear()
        notifyDataSetChanged()
    }

    fun setItemClickListener(listener: L) {
        clickListener = listener
    }
}
